lib/crypto/crypto_file.c

```diff
@@ -43,14 +45,57 @@ crypto_file_init_keys(void)
 	    256)) == NULL)
 		goto err0;
 
+	if (atexit(crypto_file_atexit)) {
+		warnp("Could not initialize atexit");
+		goto err1;
+	}
+
 	/* Success! */
 	return (0);
 
+err1:
+	rwhashtab_free(decr_aes_cache);
+
 err0:
 	/* Failure! */
 	return (-1);
 }
 
+static int
+callback_free_rwhashtab_aes_key(void * rec, void * cookie)
+{
+	struct crypto_file_aes_key * key = rec;
+
+	(void)cookie;	/* UNUSED */
+
+	/* Free expanded AES key and encrypted key lookup record. */
+	crypto_aes_key_free(key->key);
+	free(key);
+
+	/* Success! */
+	return 0;
+}
+
+/**
+ * crypto_file_atexit(void):
+ * Free the keys cached by crypto_file.
+ */
+static void crypto_file_atexit(void)
+{
+
+	/* Iterate through encrypted key -> AES key table freeing records. */
+	rwhashtab_foreach(decr_aes_cache, callback_free_rwhashtab_aes_key,
+	    NULL);
+
+	/* Free decryption cache. */
+	rwhashtab_free(decr_aes_cache);
+
+	/* Free encryption key. */
+	if (encr_aes != NULL)
+		crypto_aes_key_free(encr_aes->key);
+	free(encr_aes);
```

Comment:
free(NULL) is harmless, but it's a bit confusing since we're checking for NULL a line earlier.  I'd put this under the same if statement.  Also, I'd make the comment "Free encryption key, if we have one." to clarify why it's conditional.